api/v3/UFField.php

```diff
@@ -37,68 +37,13 @@
  * @param array $params
  *   Array per getfields metadata.
  *
- * @throws API_Exception
- *
  * @return array
  *   Newly created $ufFieldArray
  */
 function civicrm_api3_uf_field_create($params) {
-  // CRM-14756: kind of a hack-ish fix. If the user gives the id, uf_group_id is retrieved and then set.
-  if (isset($params['id'])) {
-    $groupId = civicrm_api3('UFField', 'getvalue', array(
-      'return' => 'uf_group_id',
-      'id' => $params['id'],
-    ));
-  }
-  else {
-    $groupId = CRM_Utils_Array::value('uf_group_id', $params);
-  }
-
-  $field_type       = CRM_Utils_Array::value('field_type', $params);
-  $field_name       = CRM_Utils_Array::value('field_name', $params);
-  $location_type_id = CRM_Utils_Array::value('location_type_id', $params, CRM_Utils_Array::value('website_type_id', $params));
-  $phone_type       = CRM_Utils_Array::value('phone_type_id', $params, CRM_Utils_Array::value('phone_type', $params));
-
-  if (strpos($field_name, 'formatting') !== 0 && !CRM_Core_BAO_UFField::isValidFieldName($field_name)) {
-    throw new API_Exception('The field_name is not valid');
-  }
-  $params['field_name'] = array($field_type, $field_name, $location_type_id, $phone_type);
-
-  if (!(CRM_Utils_Array::value('group_id', $params))) {
-    $params['group_id'] = $groupId;
-  }
-
-  $ids = $ufFieldArray = array();
-  $ids['uf_group'] = $groupId;
-
-  $fieldId = CRM_Utils_Array::value('id', $params);
-  if (!empty($fieldId)) {
-    $UFField = new CRM_Core_BAO_UFField();
-    $UFField->id = $fieldId;
-    if ($UFField->find(TRUE)) {
-      $ids['uf_group'] = $UFField->uf_group_id;
-      if (!(CRM_Utils_Array::value('group_id', $params))) {
-        // this copied here from previous api function - not sure if required
-        $params['group_id'] = $UFField->uf_group_id;
-      }
-    }
-    else {
-      throw new API_Exception("there is no field for this fieldId");
-    }
-    $ids['uf_field'] = $fieldId;
-  }
-
-  if (CRM_Core_BAO_UFField::duplicateField($params, $ids)) {
-    throw new API_Exception("The field was not added. It already exists in this profile.");
-  }
-  //@todo why is this even optional? Surely weight should just be 'managed' ??
-  if (CRM_Utils_Array::value('option.autoweight', $params, TRUE)) {
-    $params['weight'] = CRM_Core_BAO_UFField::autoWeight($params);
-  }
-  $ufField = CRM_Core_BAO_UFField::add($params);
+  $ufFieldArray = array();
 
-  $fieldsType = CRM_Core_BAO_UFGroup::calculateGroupType($groupId, TRUE);
-  CRM_Core_BAO_UFGroup::updateGroupTypes($groupId, $fieldsType);
+  $ufField = CRM_Core_BAO_UFField::create($params);
```

Comment:
It looks like you are now at the point where this function could be a one-liner ie. like 

https://github.com/civicrm/civicrm-core/blob/master/api/v3/Phone.php#L43

(that is the 'ideal')

The cache clearing line (104) can go in the BAO function & then the api is 'clean'